### DISCUSSION TITLE
Inverted Flat Straight Track for Hybrid Coaster

### DISCUSSION THREAD

**Discussion by Jademalo** [OpenRCT2/OpenRCT2#16588] [Feature requests]
A particularly neat RMC element is the Zero-G Stall, where the track is suspended upside down for an extended period before continuing with the inversion. The easiest example is the green track of Twisted Colossus, where it suspends itself for an extremely long time above the blue track. 

I am aware that @X123M3-256 is currently working on Zero-G rolls as a custom element, but I think a (hopefully) simple addition could allow for a lot more flexibility with stalls specifically.
I believe it could be fairly easily approximated without any new special track elements by allowing the building of inverted straight track in between barrel roll and Quarter Loop elements. 

Coasters such as the Flying Coaster have the ability to build the track in an inverted state after a half inversion allowing for sections where the guests are on their backs. If this element was added to the Hybrid coaster, it should allow for extending inversions approximating stalls. Since it's an existing element, I think this should only require the graphic and not a fully custom element.

**Comment by ocalhoun6** [comment]
Would be a great addition to other coaster types as well, particularly the Twister Coaster. (And maybe the Vertical Drop Coaster?) The sprites for it should already exist via the Flying Coaster. (Actually, come to think of it, there's every kind of inverted B&M track piece you could possibly want in the aptly named Inverted Coaster.)

(And in that case, maybe even inverted turns and s-bends from the Flying Coaster could be used as well. Imagine a half-loop that leads into two inverted turn pieces, then back into another half loop in order to make a flatter sort of cobra roll without using corkscrew pieces.)

**Comment by spacek531** [comment]
The flying coaster "inverted" track elements are actually just regular track drawn with different sprites. The trains themselves are designed for this, and the ability cannot be retroactively added (although this is less of an issue with custom rides because authors can update them). The simplest way to add an inverted flat straight track to the hybrid coaster would actually be to add a brand new inverted flat track element. Corkscrew and B&M would also benefit from this track element. Turns of any kind not be allowed because of the train design, but inverted slope pieces would be allowed.

**Reply by Jademalo** [reply]
Oh interesting, I assumed it would be just a case of creating the sprite and that the code would all be there for the most part since it's there for the flying coaster. Interesting that it's just regular straight track!

**Reply by spacek531** [reply]
The flying coaster, lay-down coaster, and multi-dimension coaster only have one car variation that the entire train uses, unlike e.g. articulated wooden which has a different front car (two variations). Flying, lay-down, and multi-dimension coasters use a second variation as the "inverted" car. The game switches the variant back and forth when the flying coaster, lay-down coaster, and multi-dimension coaster navigate one of the inverting or un-inverting track elements. The "inverted" track is just regular track and tells the game to draw the roller coaster cars at the same angles as regular track, the only difference is the car *looks* like it's inverted. Because the ride objects use pre-rendered sprites instead of runtime rendering, and because the inverted cars are saved as a second variant instead of an extension of one variant, it's very difficult to modify cars to perform like flying coaster, lay-down coaster, or multi-dimension coaster.

All this is not an issue for the feature request because it's sufficiently narrow in scope; it's some trivia why adding more inverted elements (turns and banks specifically) is not possible without invasively modifying ride objects which is not feasible.

I have thought of this track element myself, and I expect it will be relatively simple to implement.

**Reply by Jademalo** [reply]
Oh wow, that's really interesting! Thanks for the info, I really appreciate the detail.

**Reply by spacek531** [reply]
I forgot to mention one bit and that is the inversion track elements for flying coaster, lay-down coaster, and multi-dimension coaster are unique from the regular inversions. These inversions have slightly different data to facilitate switching the vehicle variant and switching the track between "inverted" and regular. Half-loops and corkscrews on the flying and lay-down will turn invisible if the track type is changed, because only those two rides have drawing code for the unique flying variant of those elements.

**Reply by Jademalo** [reply]
Oh, that makes a lot of sense then!
In my testing I swapped a flying coaster to a hybrid coaster, and noticed that the half loops turned invisible while still "working".

**Comment by X123M3-256** [comment]
I don't see the point of this. If you want a zero G stall, why not add a zero G stall? Trying to approximate it with inverted track will look terrible, and like @SpaceK says the easiest way to do this would involve adding a new track piece anyway.

**Reply by Jademalo** [reply]
After spacek's response, I agree. 
When I made the post I was under the incorrect assumption that it would be a fairly trivial code change only requiring the graphic to implement, but clearly it's just as "easy" to do a proper zero G stall element from an implementation standpoint since it would require a fully custom element anyway. 

I do still think there could be some value in inverted track pieces for some unique designs such as stalls before a quarter loop into a vertical drop, though I understand now that it's more work than just the graphic.

**Reply by X123M3-256** [reply]
>  but clearly it's just as "easy" to do a proper zero G stall element 

Well, not exactly, a proper zero G roll is much more complex to implement than an inverted straight track piece, but will give a far better result. I am intending to add zero G roll and stall elements to this ride type at some point.

**Reply by karst** [reply]
The test versions of the Zero-G roll (I imagine thisone is a special element coming from a steep hill?) that @X123M3-256 teased look incredible. Genuinly would prefer waiting for this element than converting the inline twists to flying inline twists and having upside down pieces. It just doesn't look right, way too straight for an element like this.
